### DISCUSSION TITLE
How to select the appropriate vocab for a text recognition training

### DISCUSSION THREAD

**Discussion by lfxuan** [mindee/doctr#675] [Q&A] [answered]
Hello! Why does the list (map (vocab. Index, input_string)) report an error valueerror: substring not found when training the recognition model

**Comment by charlesmindee** [comment]
Hi @lfxuan,

Thanks for reporting this, would you please give me the following:
- the code sample to reproduce the bug
- your environment (result of the `collect_env.py` script)
- eventually the image you passed to the OCR

Thanks: :pray:

**Reply by lfxuan** [reply]
Thank you for your answer!

**Comment by fg-mindee** [comment, marked as answer]
'Morning @lfxuan :wave: 

As mentioned by Charles, we would need a bit more information to have a comprehensive answer. But considering your error, I'm guessing you're training on a dataset that has characters outside of the vocab you selected :thinking: 

You can easily whether this is the case by printing the string that causes this error and then checking whether all characters are included in the vocab https://github.com/mindee/doctr/blob/main/doctr/datasets/vocabs.py (the default one on the script is "french") :+1: 

If this is the case, try to select a more appropriate vocab for your dataset, and if it doesn't exist yet in docTR, we can discuss whether we should extend the range of it :grin: 

Have a good day!

**Reply by lfxuan** [reply]
Thank you very much for your answer! The reason for the error is that the tag mistakenly adds a space when writing the JSON file. The second vocab ['french '] has no space character, which has been solved.
